renter/session.go

```diff
@@ -0,0 +1,219 @@
+package renter
+
+import (
+	"crypto/ed25519"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/net/rhp"
+	"go.sia.tech/core/net/rpc"
+	"go.sia.tech/core/types"
+)
+
+type (
+	// ChainManager manages the chain state.
+	ChainManager interface {
+		// Tip returns the tip of the best known valid chain.
+		Tip() types.ChainIndex
+		// ValidationContext returns the ValidationContext for the specified
+		// index.
+		ValidationContext(index types.ChainIndex) (consensus.ValidationContext, error)
+	}
+
+	// A TransactionPool broadcasts transaction sets to miners for inclusion in an
+	// upcoming block.
+	TransactionPool interface {
+		AcceptTransactionSet(txns []types.Transaction) error
+		FeeEstimate() (min, max types.Currency, err error)
+		UnconfirmedParents(txn types.Transaction) ([]types.Transaction, error)
+		BroadcastTransaction(txn types.Transaction, dependsOn []types.Transaction)
+	}
+
+	// A Wallet provides addresses and funds and signs transactions.
+	Wallet interface {
+		Balance() types.Currency
+		Address() types.Address
+		Addresses() []types.Address
+		FundTransaction(txn *types.Transaction, amount types.Currency, pool []types.Transaction) ([]types.ElementID, func(), error)
+		SignTransaction(vc consensus.ValidationContext, txn *types.Transaction, toSign []types.ElementID) error
+	}
+)
+
+// A Session is an implementation of the renter side of the renter-host protocol.
+type Session struct {
+	secret  ed25519.PrivateKey
+	hostKey types.PublicKey
+
+	wallet  Wallet
+	tpool   TransactionPool
+	manager ChainManager
+	session *rhp.Session
+
+	settingsMu         sync.RWMutex
+	settings           types.HostSettings
+	settingsExpiration time.Time
+}
+
+func (s *Session) currentSettings() (types.HostSettings, error) {
+	s.settingsMu.RLock()
+	defer s.settingsMu.RUnlock()
+	if s.settingsExpiration.Before(time.Now()) {
```

Comment:
the eternal question: which is more readable, `expiration.Before(time.Now())`, or `time.Now().After(expiration)`? :P